### PR TITLE
Fix #780 : Added top/bottom for Tabs Bar in Tab Panel UI

### LINKED ISSUE #780: Tabs bar positioning in TabUI
**Describe the solution you'd like**

The default behavior of tab-like UI components usually places them right above the content, e.g., web browser tabs. However, it would be helpful if, during the creation of a TabUI element, the user could specify where (top, bottom, left, or right) the tab bar should be placed.

tab_layout

### PR DESCRIPTION
**Pull request by IshanKamboj**
# Overview
1. This PR fixes or fixes part of #780 
2. This PR does the following: Added `tabBar_pos` as a new parameter in TabUI class to specify position of the tabs bar in the Tabs Panel UI. The default value of this parameter is set as top. Currently this parameter accept `top` and `bottom` as inputs. Further PR will also add left/right inputs. This is done in order to make Tabs Panel UI more customisable.

# Proof that changes work
### Top
image

image

### Bottom
image

image

### CODE REVIEW

**Comment by IshanKamboj** [comment]
@guaje PTAL 
Thank you

**Comment by ganimtron-10** [comment]
Hey @IshanKamboj,
Thanks for doing this!
Could you also please add the relevant tests for the changes?

**Comment by IshanKamboj** [comment]
> Hey @IshanKamboj, Thanks for doing this! Could you also please add the relevant tests for the changes?

okay ... will update after adding the relevant tests

**Comment by IshanKamboj** [comment]
@ganimtron-10 PTAL .... I added the test that I think is relevant. I separated the Tabs UI test into two separate parts for top and bottom of Tabs Bar.
Thank you

File: fury/ui/tests/test_containers.py [hunk 1]
```diff
@@ -360,3 +361,117 @@ def tab_change(tab_ui):
...
+        print(list(event_counter.events_counts.items()))
+        event_counter.save(expected_events_counts_filename)
+    else:
+        show_manager.play_events_from_file(recording_filename)
+        expected = EventCounter.load(expected_events_counts_filename)
+        event_counter.check_counts(expected)
+
+    npt.assert_equal(0, tab_ui.active_tab_idx)
+    npt.assert_equal(11, next(changes))
+    npt.assert_equal(4, next(collapses))
```
Comment (guaje) [inline fury/ui/tests/test_containers.py:477]: Although I understand your motivation, try to include both the `top` and `bottom` tests in the `test_ui_tab_ui`. If you need a specific test for each, then please make sure to make it specific, since this solution duplicates code and complicates maintenance in the long term.

**Comment by guaje** [comment]
I have enabled the CI tests for this PR. Please address the respective comments and issues.

**Comment by ganimtron-10** [comment]
> @ganimtron-10 PTAL .... I added the test that I think is relevant. I separated the Tabs UI test into two separate parts for top and bottom of Tabs Bar.
> Thank you

Thanks for the update. I will be able to look into it on Saturday meanwhile please make sure to resolve all the above comments.

File: fury/ui/tests/test_containers.py (hunk 1 above)
Comment (IshanKamboj) [reply fury/ui/tests/test_containers.py:477]: sure will try to do make a combined test for both

**Comment by IshanKamboj** [comment]
@guaje , I combined the tests in the same function `test_ui_tab_ui` using a for loop, Is that acceptable? or should I do something else? ... I did this as I think that the test for the Tabs UI would remain the same and just the position of Tabs Bar will be changing and show that it still works fine.
Thank you

File: fury/ui/tests/test_containers.py [hunk 2]
```diff
@@ -250,113 +250,118 @@ def test_ui_image_container_2d(interactive=False):
 
 
 def test_ui_tab_ui(interactive=False):
-    filename = 'test_ui_tab_ui'
-    recording_filename = pjoin(DATA_DIR, filename + '.log.gz')
-    expected_events_counts_filename = pjoin(DATA_DIR, filename + '.json')
+    positions = ['top', 'bottom']
+    for i in positions:
```
Comment (ganimtron-10) [inline fury/ui/tests/test_containers.py:254]: I doubt using for loop may cause few issues moving forward. It would make it difficult to understand in which loop did the test failed whether it was the `top` case or `bottom` case.
Please have a look at other tests where we have checked multiple parameters in a single test using multiple instance of same UI element.

File: fury/ui/containers.py [hunk 3]
```diff
@@ -619,6 +619,7 @@ def __init__(
         inactive_color=(0.5, 0.5, 0.5),
         draggable=False,
         startup_tab_id=None,
+        tabBar_pos="top",
```
Comment (ganimtron-10) [inline fury/ui/containers.py:622]: Please follow `snake_case` while naming the variable.
Something like `tab_bar_pos` `tab_header_pos` or `tab_panel_pos`

File: fury/ui/tests/test_containers.py [hunk 4]
```diff
@@ -250,113 +250,118 @@ def test_ui_image_container_2d(interactive=False):
...
+        npt.assert_equal(tab_ui.tabs[1].title_font_size, 18)
+        npt.assert_equal(tab_ui.tabs[2].title_font_size, 18)
+
+        tab_ui.tabs[0].title_font_size = 10
+        tab_ui.tabs[1].title_font_size = 20
+        tab_ui.tabs[2].title_font_size = 30
+
+        npt.assert_equal(tab_ui.tabs[0].title_font_size, 10)
+        npt.assert_equal(tab_ui.tabs[1].title_font_size, 20)
+        npt.assert_equal(tab_ui.tabs[2].title_font_size, 30)
```
Comment (ganimtron-10) [inline fury/ui/tests/test_containers.py:301]: These tests are not required again. Update the unit test to check the functionality you have added as this is repetition of the tests which are already being checked.

**Review by ganimtron-10 — COMMENTED** [review]
Hey @IshanKamboj ,
Thanks for the updates.
It would be great if we keep the scope of a PR restricted to an individual task mentioned in it as in longer run if we try to combine multiple things in a same PR it will be difficult to manage and revert things back.

So please separate the doc typo fixes and the sprite test resolution to individual PRs so its easier to manage.

Also PTAL on below comments.

Thanks!

File: fury/tests/test_io.py [hunk 5]
```diff
@@ -241,8 +241,8 @@ def test_load_cubemap_texture():
 
 def test_load_sprite_sheet():
     sprite_URL = (
-        'https://raw.githubusercontent.com/'
-        'antrikshmisri/DATA/master/fury/0yKFTBQ.png'
+        'https://raw.githubusercontent.com/IshanKamboj/'
+        'IshanKamboj/main/sprites.png'
```
[the remark this answers is not in the document]
Comment (IshanKamboj) [reply fury/tests/test_io.py:245]: Ok I m sorry to do this in the same PR... The CI checks were failing So I thought of fixing them but didn't realise it should be addressed as a seprate issue altogether.

File: fury/ui/tests/test_containers.py (hunk 2 above)
Comment (IshanKamboj) [reply fury/ui/tests/test_containers.py:254]: Got it I will try doing it the other way
Thank you

**Comment by IshanKamboj** [comment]
@ganimtron-10 PTAL ... I am really sorry I mistakenly did a force push. Although I did the changes that you requested.

File: fury/ui/tests/test_containers.py (hunk 2 above)
Comment (ganimtron-10) [reply fury/ui/tests/test_containers.py]: As far as I remember we have kept the same instance of show manager and other things like event counter and elements but just created the UI object with different parameters for testing. Let me see your previous test and will comment on it.

File: fury/ui/containers.py [hunk 6]
```diff
@@ -640,6 +641,8 @@ def __init__(
         startup_tab_id : int, optional
             Tab to be activated and uncollapsed on startup.
             by default None is activated/ all collapsed.
+        tabBar_pos : str, optional
```
Comment (ganimtron-10) [inline fury/ui/containers.py:644]: documentation is not updated

File: fury/ui/containers.py [hunk 7]
```diff
@@ -640,6 +641,8 @@ def __init__(
...
+            Position of the Tabs Bar in the panel
```
Comment (ganimtron-10) [inline fury/ui/containers.py:645]: `tab` instead of `tabs`

File: fury/ui/containers.py [hunk 8]
```diff
@@ -715,8 +719,10 @@ def _get_size(self):
     def update_tabs(self):
         """Update position, size and callbacks for tab panels."""
         self.tab_panel_size = (self.size[0] // self.nb_tabs, int(0.1 * self.size[1]))
-
-        tab_panel_pos = [0.0, 0.9]
+        if self.tab_bar_pos == "top":
```
Comment (ganimtron-10) [inline fury/ui/containers.py:722]: `self.tab_bar_pos.lower()` instead

File: fury/ui/containers.py [hunk 9]
```diff
@@ -715,8 +719,10 @@ def _get_size(self):
...
+            tab_panel_pos = [0.0, 0.9]
+        elif self.tab_bar_pos == "bottom":
```
Comment (ganimtron-10) [inline fury/ui/containers.py:724]: same as above

File: fury/ui/containers.py [hunk 10]
```diff
@@ -754,7 +760,12 @@ def update_tabs(self):
 
             tab_panel.content_panel.resize(self.content_size)
             self.parent_panel.add_element(tab_panel, tab_panel_pos)
-            self.parent_panel.add_element(tab_panel.content_panel, (0.0, 0.0))
+            if self.tab_bar_pos == "top":
```
Comment (ganimtron-10) [inline fury/ui/containers.py:763]: same as above

File: fury/ui/containers.py [hunk 11]
```diff
@@ -754,7 +760,12 @@ def update_tabs(self):
...
+                self.parent_panel.add_element(tab_panel.content_panel,
+                                              (0.0, 0.0))
+            elif self.tab_bar_pos == "bottom":
```
Comment (ganimtron-10) [inline fury/ui/containers.py:766]: same as above

**Review by ganimtron-10 — COMMENTED** [review]
Hey @IshanKamboj ,

The test still needs some update, instead of separating the sections into two part and repeating the whole flow just repeat the elements that you are testing
ie.
1. It should have a common recording file, event counter, show_manager
2. Just create two different instance of `TabUI` one with `top` and one with `bottom` and then test the things which are affected due to this change and you will be good to go.

Also look at the below comments.

Thanks in advance!

**Comment by IshanKamboj** [comment]
Hello @ganimtron-10 PTAL I changed the Tab UI test again, now I am only checking the thing affected by changing the position of Tab Bar from top to bottom. As essentially the position of the Tab Bar is changing with respect to the content panel it should only affect the elements in the content panel. So now I am just testing for the changes(addition,removal and update) in the elements for the bottom part. Also as other functionality like callbacks is not affected we need not check the same for the bottom part.

Thank you

**Comment by IshanKamboj** [comment]
Hey @ganimtron-10 just a quick reminder can you please take a look at my previous comment and changes that I made and give a feedback on if the test is okay or does it still require some other changes.

Thank you

**Comment by ganimtron-10** [comment]
Hey @IshanKamboj ,
Thanks for reminder. I have planned it to do by EOD.

**Comment by ganimtron-10** [comment]
Hey @IshanKamboj ,
I think you misunderstood about the tests. Below is a simplified explanation of how we are expecting test.

For now we have the test called `test_ui_tab_ui` which basically test the basic functionalities of the TabUI. As this is a new addition of feature it is better to have a separate function ie. `test_ui_tab_ui_tab_position`. So, this new function would have the below kind of format. You can refer this [test1](https://github.com/fury-gl/fury/blob/2a35d023138b99cb336247d3e893780f9a55e6ac/fury/ui/tests/test_core.py#L387) or [test2](https://github.com/fury-gl/fury/blob/2a35d023138b99cb336247d3e893780f9a55e6ac/fury/ui/tests/test_elements.py#L272) for more context.

```python
def test_ui_tab_ui_tab_position(interactive=False):
    filename = 'test_ui_tab_ui_tab_position'
    recording_filename = pjoin(DATA_DIR, filename + '.log.gz')
    expected_events_counts_filename = pjoin(DATA_DIR, filename + '.json')

    tab_ui_top = ui.TabUI(
        position=(50, 50), size=(300, 300), nb_tabs=3, draggable=True,
        tab_bar_pos='top')


    tab_ui_bottom = ui.TabUI(
        position=(50, 50), size=(300, 300), nb_tabs=3, draggable=True,
        tab_bar_pos='bottom'
    )

    # test and assert various properties, positions and changes which is affected by the change
  
    # Create Event counter for both tab_ui_top and tab_ui_bottom

    # Create a single scene for both the tab
    # Add both the tabs to the scene

    if interactive:
        # manage and record events for both tab ui's
        # while you are recording test try collapsing various tabs, changing tabs and moving them around for both tab ui
    else:
        # replay events for both tab ui's

    # check the collapse, and changes count

```

**Comment by IshanKamboj** [comment]
> Hey @IshanKamboj , I think you misunderstood about the tests. Below is a simplified explanation of how we are expecting test.
> 
> For now we have the test called `test_ui_tab_ui` which basically test the basic functionalities of the TabUI. As this is a new addition of feature it is better to have a separate function ie. `test_ui_tab_ui_tab_position`. So, this new function would have the below kind of format. You can refer this [test1](https://github.com/fury-gl/fury/blob/2a35d023138b99cb336247d3e893780f9a55e6ac/fury/ui/tests/test_core.py#L387) or [test2](https://github.com/fury-gl/fury/blob/2a35d023138b99cb336247d3e893780f9a55e6ac/fury/ui/tests/test_elements.py#L272) for more context.
> 
> ```python
> def test_ui_tab_ui_tab_position(interactive=False):
>     filename = 'test_ui_tab_ui_tab_position'
>     recording_filename = pjoin(DATA_DIR, filename + '.log.gz')
>     expected_events_counts_filename = pjoin(DATA_DIR, filename + '.json')
> 
>     tab_ui_top = ui.TabUI(
>         position=(50, 50), size=(300, 300), nb_tabs=3, draggable=True,
>         tab_bar_pos='top')
> 
> 
>     tab_ui_bottom = ui.TabUI(
>         position=(50, 50), size=(300, 300), nb_tabs=3, draggable=True,
>         tab_bar_pos='bottom'
>     )
> 
>     # test and assert various properties, positions and changes which is affected by the change
>   
>     # Create Event counter for both tab_ui_top and tab_ui_bottom
> 
>     # Create a single scene for both the tab
>     # Add both the tabs to the scene
> 
>     if interactive:
>         # manage and record events for both tab ui's
>         # while you are recording test try collapsing various tabs, changing tabs and moving them around for both tab ui
>     else:
>         # replay events for both tab ui's
> 
>     # check the collapse, and changes count
> ```

Understood now 👍 , will do it this way. Sorry for the confusion above
Thank you

**Comment by IshanKamboj** [comment]
@ganimtron-10 PTAL now, I added the test in the way you explained. Please tell if there are still any issues with it.
Thank you

**Comment by IshanKamboj** [comment]
Hey @ganimtron-10 PTAL I added the test in the way you asked. If there is still some issues please let me know.
Thank you

**Comment by ganimtron-10** [comment]
Hey @IshanKamboj , This looks good to me.
@skoudoro @guaje PTAL.

**Comment by skoudoro** [comment]
I will look into it this week. Thank you the review @ganimtron-10  and thank you for the work @IshanKamboj

**Comment by skoudoro** [comment]
Also, can you update `viz_tab` tutorial to show this features @IshanKamboj. thank you

**Comment by IshanKamboj** [comment]
Hey @skoudoro PTAL, I updated the viz_tab tutorial to include this feature.
Sorry for the delay
Thank you

**Comment by IshanKamboj** [comment]
Hey @skoudoro just a quick reminder PTAL.
Thank you

**Comment by skoudoro** [comment]
@maharshi-gor, Can you try and look at this?  I do not find the time to do a deeper test and look in details.

Thank you in advance.

@IshanKamboj, sorry for the delay

File: fury/ui/containers.py [hunk 12]
```diff
@@ -737,8 +740,10 @@ def _get_size(self):
     def update_tabs(self):
         """Update position, size and callbacks for tab panels."""
         self.tab_panel_size = (self.size[0] // self.nb_tabs, int(0.1 * self.size[1]))
-
-        tab_panel_pos = [0.0, 0.9]
+        if self.tab_bar_pos.lower() == "top":
+            tab_panel_pos = [0.0, 0.9]
+        elif self.tab_bar_pos.lower() == "bottom":
+            tab_panel_pos = [0.0, 0.0]
```
Comment (skoudoro) [inline fury/ui/containers.py:746]: What happen if it is not "top" or "bottom" for `self.tab_bar_pos`

File: fury/ui/containers.py [hunk 13]
```diff
@@ -776,7 +781,12 @@ def update_tabs(self):
 
             tab_panel.content_panel.resize(self.content_size)
             self.parent_panel.add_element(tab_panel, tab_panel_pos)
-            self.parent_panel.add_element(tab_panel.content_panel, (0.0, 0.0))
+            if self.tab_bar_pos.lower() == "top":
+                self.parent_panel.add_element(tab_panel.content_panel,
+                                              (0.0, 0.0))
+            elif self.tab_bar_pos.lower() == "bottom":
+                self.parent_panel.add_element(tab_panel.content_panel,
+                                              (0.0, 0.1))
```
Comment (skoudoro) [inline fury/ui/containers.py:789]: same question as above ?

**Review by skoudoro — COMMENTED** [review]
Hi @IshanKamboj,

Apologize for the late review and thank you for doing this. See below my question. 

Thank you for the feedback

File: fury/ui/containers.py (hunk 12 above)
Comment (IshanKamboj) [reply fury/ui/containers.py:746]: Ohh Sorry my bad... I missed that. In case if it is not 'top' or 'bottom' I should throw a value error

File: fury/ui/containers.py (hunk 13 above)
Comment (IshanKamboj) [reply fury/ui/containers.py:789]: raising ValueError in the above part should take care of this also.

**Comment by IshanKamboj** [comment]
@skoudoro PTAL, I added the warning. Should I use logging or is this ok?

File: fury/ui/containers.py (hunk 12 above)
Comment (skoudoro) [reply fury/ui/containers.py:750]: you should use warnings module

`from warning import warn.` then `warn(my_msg)`

**Review by skoudoro — COMMENTED** [review]
thank you for the update @IshanKamboj,

See below my comment

File: fury/ui/containers.py (hunk 13 above)
Comment (skoudoro) [reply fury/ui/containers.py:794]: ok, so do not forget to take care of this

**Comment by IshanKamboj** [comment]
@skoudoro I added the warning as u specified..... also I only need to check `self.tab_bar_pos` is top/bottom at the beginning of the function and if it is not then I just set it `top` by default so both the conditions below that should work.

**Comment by IshanKamboj** [comment]
@skoudoro just a quick reminder please review the changes I made... 
Thank you

**Comment by skoudoro** [comment]
Waiting for the CI's to run and then I will go ahead and merge this PR. It seems ok. Thank you for this @IshanKamboj.

**Comment by skoudoro** [comment]
Thank you for this work @IshanKamboj !
